osu.Game/Rulesets/Difficulty/Skills/StrainSkill.cs

```diff
@@ -0,0 +1,133 @@
+
+        private double currentSectionEnd;
+
+        private readonly List<double> strainPeaks = new List<double>();
+
+        /// <summary>
+        /// Process a <see cref="DifficultyHitObject"/> and update current strain values accordingly.
+        /// </summary>
+        public sealed override void Process(DifficultyHitObject current)
+        {
+            // The first object doesn't generate a strain, so we begin with an incremented section end
+            if (Previous.Count == 0)
+                currentSectionEnd = Math.Ceiling(current.StartTime / SectionLength) * SectionLength;
+
+            while (current.StartTime > currentSectionEnd)
+            {
+                saveCurrentPeak();
+                startNewSectionFrom(currentSectionEnd);
+                currentSectionEnd += SectionLength;
+            }
+
+            CurrentStrain *= strainDecay(current.DeltaTime);
+            CurrentStrain += StrainValueOf(current) * SkillMultiplier;
+
+            currentSectionPeak = Math.Max(CurrentStrain, currentSectionPeak);
+
+            base.Process(current);
+        }
+
+        /// <summary>
+        /// Saves the current peak strain level to the list of strain peaks, which will be used to calculate an overall difficulty.
+        /// </summary>
+        private void saveCurrentPeak()
+        {
+            strainPeaks.Add(currentSectionPeak);
+        }
+
+        /// <summary>
+        /// Sets the initial strain level for a new section.
+        /// </summary>
+        /// <param name="time">The beginning of the new section in milliseconds.</param>
+        private void startNewSectionFrom(double time)
+        {
+            // The maximum strain of the new section is not zero by default, strain decays as usual regardless of section boundaries.
+            // This means we need to capture the strain level at the beginning of the new section, and use that as the initial peak level.
+            currentSectionPeak = GetPeakStrain(time);
+        }
+
+        /// <summary>
+        /// Retrieves the peak strain at a point in time.
+        /// </summary>
+        /// <param name="time">The time to retrieve the peak strain at.</param>
+        /// <returns>The peak strain.</returns>
+        protected virtual double GetPeakStrain(double time) => CurrentStrain * strainDecay(time - Previous[0].StartTime);
+
+        /// <summary>
+        /// Returns the calculated difficulty value representing all processed <see cref="DifficultyHitObject"/>s.
+        /// </summary>
+        public sealed override double DifficultyValue()
+        {
+            double difficulty = 0;
+            double weight = 1;
+
+            var finalStrainPeaks = strainPeaks.Select(p => p).ToList(); // Copy list so we don't create side effects
+            finalStrainPeaks.Add(currentSectionPeak); // Add final peak since it never gets saved
```

Comment:
This is sort of dodgy. Ideally `Process()` would be doing this instead of having this cleanup in another method that looks to be unrelated.

This is the consequence of the choice of having `Process()` take in individual HOs instead of the entire list. Passing the whole list down would avoid this problem, I think.

Reply:
This exists to ensure that `DifficultyValue()` is kept as a pure function. It merely interprets the current state of the `Skill` and gives you a result. If we were to pass the entire list of hitobjects to the `Skill` in a single process call, then in that case why even let it have state, static would work too? why have a `DifficultyValue()` method? we could just return the difficulty value as the result of the `ProcessAll()` method.
This is certainly a valid approach, but I believe it's not the correct one. Correct me if I'm wrong but, I believe we made the conscious decision to have `Skill`s follow a stateful approach rather than a static one for this very reason.

Consider if we wanted to get the difficulty of a `Skill` after processing each object for some reason (perhaps constructing a difficulty graph).
If we were to add this incomplete strain to `strainPeaks`, the following `Process()` calls would result in incorrect values.
I think we should be able to call `Process()` `Process()` `DifficultyValue()` and end up with the same result as `Process()` `DifficultyValue()` `Process()` `DifficultyValue()`.

Reply:
I misunderstood what this was doing, partially because of the xmldoc that states the value represents "all processed difficulty hitobjects". I thought it was a hack that works around the fact that `Process()` doesn't add the final section strain (because it doesn't actually know when it receives the final difficulty hit object). Maybe the xmldoc of this method should state "...up to now" or similar to clarify.

Reply:
Yea that's understandable. I'll make it a bit more clear in the xmldoc.
FIxed in 79a6222a23e828d59a1cfeb34cdb9695d40758dc